modules/compiler/utils/source/barrier_regions.cpp

```diff
@@ -879,23 +923,33 @@ void compiler::utils::Barrier::MakeLiveVariableMemType() {
       member_ty = field_ty;
     }
 
-    // For a scalable vector, we need the size of the equivalent fixed vector
-    // based on its known minimum size.
-    auto member_ty_fixed = member_ty;
-    if (isa<ScalableVectorType>(member_ty)) {
-      auto *const eltTy = multi_llvm::getVectorElementType(member_ty);
-      auto n = multi_llvm::getVectorElementCount(member_ty).getKnownMinValue();
-      member_ty_fixed = VectorType::get(eltTy, ElementCount::getFixed(n));
+    std::vector<Type *> member_tys = {member_ty};
+    // If this is a struct type containing any scalable members, we must
+    // decompose the value into its individual components.
+    if (isStructWithScalables(member_ty)) {
+      member_tys = cast<StructType>(member_ty)->elements().vec();
     }
 
-    // Need to ensure that alloc alignment or preferred alignment is kept
-    // in the new struct so pad as necessary.
-    unsigned size = dl.getTypeAllocSize(member_ty_fixed);
-    alignment = std::max(dl.getPrefTypeAlign(member_ty).value(),
-                         static_cast<AlignIntTy>(alignment));
-    max_live_var_alignment = std::max(alignment, max_live_var_alignment);
+    for (auto [idx, ty] : enumerate(member_tys)) {
+      // For a scalable vector, we need the size of the equivalent fixed vector
+      // based on its known minimum size.
+      auto member_ty_fixed = ty;
+      if (isa<ScalableVectorType>(ty)) {
+        auto *const eltTy = multi_llvm::getVectorElementType(ty);
+        auto n = multi_llvm::getVectorElementCount(ty).getKnownMinValue();
+        member_ty_fixed = VectorType::get(eltTy, ElementCount::getFixed(n));
+      }
 
-    barrier_members.push_back({i, member_ty, alignment, size});
+      // Need to ensure that alloc alignment or preferred alignment is kept
+      // in the new struct so pad as necessary.
+      unsigned size = dl.getTypeAllocSize(member_ty_fixed);
+      alignment = std::max(dl.getPrefTypeAlign(ty).value(),
+                           static_cast<AlignIntTy>(alignment));
+      max_live_var_alignment = std::max(alignment, max_live_var_alignment);
+
+      barrier_members.push_back(
+          {i, static_cast<unsigned>(idx), ty, alignment, size});
```

Comment:
since the loop is a bit  bigger, maybe a good time to rename `i` to something like `live_var`

Reply:
Good idea, thanks